### PR TITLE
For #11100, change 'add reource' syntax and  allow custom properties for connection pool.

### PR DESCRIPTION
**Pull request by RaigorJiang**
For #11100.

Changes proposed in this pull request:
- Remove `HOST`, `PORT`, `DBNAME` and add `URL` to `ADD RESOURCE` syntax.
- Users can use custom url to set JDBC connection properties.
- User can set properties to customize  the connection pool, just like `maxPoolSize`;

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java [hunk 1]
```diff
@@ -27,23 +28,33 @@
...
+        addJDBCProperty(dataSource, "useLocalSessionState", Boolean.TRUE.toString());
+        addJDBCProperty(dataSource, "rewriteBatchedStatements", Boolean.TRUE.toString());
+        addJDBCProperty(dataSource, "cacheResultSetMetadata", Boolean.FALSE.toString());
+        addJDBCProperty(dataSource, "cacheServerConfiguration", Boolean.TRUE.toString());
+        addJDBCProperty(dataSource, "elideSetAutoCommits", Boolean.TRUE.toString());
+        addJDBCProperty(dataSource, "maintainTimeStats", Boolean.FALSE.toString());
+        addJDBCProperty(dataSource, "netTimeoutForStreamingResults", "0");
+        addJDBCProperty(dataSource, "tinyInt1isBit", Boolean.FALSE.toString());
+        addJDBCProperty(dataSource, "useSSL", Boolean.FALSE.toString());
+        addJDBCProperty(dataSource, "serverTimezone", "UTC");
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java:45]: I read some code of MySQL JDBC Driver 5.1.47. There is a doc for `com.mysql.jdbc.NonRegisteringDriver#connect`wrote
>  The java.util.Properties argument can be used to pass arbitrary string tag/value pairs as connection arguments. These properties take precedence over any properties sent in the URL.

My understanding is that the parameters such as `useSSL=true` we write in the URL will not work.

Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java:45]: Thanks @TeslaCN .
I have read `parseURL` method of class `NonRegisteringDriver` too.  So I add a 'url check' in `addJDBCProperty()`, please have a see.

**Comment by RaigorJiang** [comment]
@tristaZero  Now users can add resource with :
```
ADD RESOURCE 
resource_0 (
    HOST=127.0.0.1,
    PORT=3306,
    DB=db0,
    USER=root,
    PASSWORD=root,
    PROPERTIES("maxPoolSize"=50, "idleTimeoutMilliseconds"="60000")
    ), 
resource_1 (
    URL="jdbc:mysql://127.0.0.1:3306/db1?serverTimezone=UTC&useSSL=false",
    USER=root,
    PASSWORD=root,
    PROPERTIES("maxPoolSize"=50, "idleTimeoutMilliseconds"="60000")
    );
```

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java (hunk 1 above)
Comment (TeslaCN) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java:45]: I think using `indexOf` is too risk. Is there a better way?

Comment (TeslaCN) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/datasource/decorator/HikariJDBCParameterDecorator.java:45]: How about using `java.net.URI` to get the query part?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Get properties map of JDBC url.
+     *
+     * @return properties map
+     */
+    public Map<String, String> getQueryMap() {
+        if (getQuery().isPresent()) {
+            return Splitter.on("&").withKeyValueSeparator("=").split(getQuery().get());
+        }
+        return new HashMap<>(16, 1);
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:92]: Consider using `emptyMap()`.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    public ConnectionUrlParser(final String jdbcUrl) {
+        matcher = CONNECTION_URL_PATTERN.matcher(jdbcUrl);
+        if (!matcher.matches()) {
+            throw new ShardingSphereConfigurationException("Incorrect JDBC url format: %s", jdbcUrl);
+        }
+        this.scheme = matcher.group(KEY_SCHEME);
+        this.authority = matcher.group(KEY_AUTHORITY);
+        this.path = matcher.group(KEY_PATH);
+        this.query = matcher.group(KEY_QUERY);
+    }
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:72]: Consider removing `this.`.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    
+    private final Matcher matcher;
+    
+    private String scheme;
+    
+    private String authority;
+    
+    private String path;
+    
+    private String query;
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:61]: Can these fields be final?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * JDBC connection url parser.
+ */
+public final class ConnectionUrlParser {
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:33]: It is better to test this class with variant JDBC URL.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java (hunk 2 above)
Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:92]: This param Map is uesd to put item, so emptyMap() dose not work.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java (hunk 3 above)
Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:72]: ok

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java (hunk 4 above)
Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:61]: yes, good advice.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java (hunk 5 above)
Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:33]: ok

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java (hunk 2 above)
Comment (TeslaCN) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/connection/ConnectionUrlParser.java:92]: I didn't find where will put KV into the map, only saw the `isEmpty` and `containsKey`. Could you show me where?

**Review by TeslaCN — APPROVED** [review]
LGTM

**Pull request merged**